app/src/main/java/com/ataulm/whatsnext/api/LetterboxdApiFactory.kt

```diff
@@ -0,0 +1,105 @@
+package com.ataulm.whatsnext.api
+
+import com.ataulm.support.Clock
+import com.ataulm.whatsnext.TokensStore
+import okhttp3.*
+import okhttp3.logging.HttpLoggingInterceptor
+import retrofit2.Retrofit
+import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
+import retrofit2.converter.gson.GsonConverterFactory
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+class LetterboxdApiFactory(
+        private val apiKey: String,
+        private val apiSecret: String,
+        private val tokensStore: TokensStore,
+        private val clock: Clock,
+        private val enableHttpLogging: Boolean
+) {
+
+    fun createRemote(): LetterboxdApi {
+        val okHttpClient = OkHttpClient.Builder()
+                .addNetworkInterceptor(AddApiKeyQueryParameterInterceptor(apiKey, clock))
+                .addNetworkInterceptor(AddAuthorizationHeaderInterceptor(apiSecret, tokensStore))
+
+        if (enableHttpLogging) {
+            val httpLoggingInterceptor = HttpLoggingInterceptor()
+            httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
+            okHttpClient.addInterceptor(httpLoggingInterceptor)
+        }
+
+        return Retrofit.Builder()
+                .client(okHttpClient.build())
+                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
+                .addConverterFactory(GsonConverterFactory.create())
+                .baseUrl("https://api.letterboxd.com/api/v0/")
+                .build()
+                .create(LetterboxdApi::class.java)
+    }
+}
+
+private class AddApiKeyQueryParameterInterceptor(private val apiKey: String, private val clock: Clock) : Interceptor {
+
+    override fun intercept(chain: Interceptor.Chain): Response {
+        val url = chain.request().url.newBuilder()
+                .addQueryParameter("apikey", apiKey)
+                .addQueryParameter("nonce", UUID.randomUUID().toString())
+                .addQueryParameter("timestamp", TimeUnit.MILLISECONDS.toSeconds(clock.currentTimeMillis).toString())
+                .build()
+        val request = chain.request().newBuilder().url(url).build()
+        return chain.proceed(request)
+    }
+}
+
+private const val HEADER_AUTH = "Authorization"
+
+private class AddAuthorizationHeaderInterceptor(private val apiSecret: String, private val tokensStore: TokensStore) : Interceptor {
```

Comment:
This is the oddest API I've seen yet. If I didn't have an example client in ruby, I'd never have been able to get this working.